dockerscripts/healthcheck.go

```diff
@@ -0,0 +1,122 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+const (
+	initGraceTime = 300
+	healthPath    = "/minio/health/live"
+	scheme        = "http://"
+	schemeSecure  = "https://"
+)
+
+// returns container boot time by finding
+// modtime of /proc/1 directory
+func getStartTime() time.Time {
+	di, err := os.Stat("/proc/1")
+	if err != nil {
+		// Cant stat proc dir successfully, exit with error
+		os.Exit(1)
+	}
+	return di.ModTime()
+}
+
+// Finds the ip/port of the Minio process
+// running in the container
+func findEndpoint() string {
+	op, err := exec.Command("netstat", "-ntlp").Output()
+	if err != nil {
+		// netstat unsuccessful, exit with error
+		os.Exit(1)
+	}
+	// split netstat output in rows
+	lines := strings.Split(string(op), "\n")
+	// loop over the rows to find Minio process
+	for _, line := range lines {
+		if strings.Contains(line, "minio") {
+			newLine := strings.Replace(line, ":::", "127.0.0.1:", 1)
+			fields := strings.Fields(newLine)
+			// index 3 in the row has the Local address
+			// find the last index of ":" - address will
+			// have port number after this index
+			i := strings.LastIndex(fields[3], ":")
+			// split address and port
+			addr := fields[3][:i]
+			port := fields[3][i+1:]
+			// add surrounding [] for ip6 address
+			if strings.Count(addr, ":") > 0 {
+				addr = strings.Join([]string{"[", addr, "]"}, "")
+			}
+			// return joint address and port
+			return strings.Join([]string{addr, port}, ":")
+		}
+	}
+	// minio process not found, exit with error
+	os.Exit(1)
+	return ""
+}
+
+func main() {
+	startTime := getStartTime()
+
+	//  In distributed environment like Swarm, traffic is routed
+	//  to a container only when it reports a `healthy` status. So, we exit
+	//  with 0 to ensure healthy status till distributed Minio starts (120s).
+
+	//  Refer: https://github.com/moby/moby/pull/28938#issuecomment-301753272
+
+	if (time.Now().Sub(startTime) / time.Second) < initGraceTime {
+		os.Exit(0)
+	} else {
+		endPoint := findEndpoint()
+		uri := strings.Join([]string{scheme, endPoint, healthPath}, "")
+		resp, err := http.Get(uri)
+		if err != nil {
+			// GET failed exit
+			os.Exit(1)
+		}
+		if resp.StatusCode == http.StatusOK {
+			// exit with success
+			os.Exit(0)
+		}
+		bodyBytes, err := ioutil.ReadAll(resp.Body)
+		bodyString := string(bodyBytes)
+		resp.Body.Close()
+		// This means sever is configured with https
+		if resp.StatusCode == http.StatusForbidden && bodyString == "SSL required" {
+			// Try with https scheme
+			uri = strings.Join([]string{schemeSecure, endPoint, healthPath}, "")
+			resp, err = http.Get(uri)
```

Comment:
We probably need to find a way to make this works in case of self signed certificate (although a workaround for the user is to install the certificate in the OS)